### PR TITLE
[Enhancement] Support MultiScaleDeformableAttention with AMP

### PR DESCRIPTION
**Pull request by nijkah**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Currently, `MultiScaleDeformableAttention` with AMP is not supported.
```
RuntimeError: "ms_deform_attn_forward_cuda" not implemented for 'Half'
```

## Modification
Following https://discuss.pytorch.org/t/how-can-i-write-the-cuda-code-to-support-fp16-calculation/107181 and `DeformConv2d`,
apply `AT_DISPATCH_FLOATING_TYPES_AND_HALF` and solve the type mismatch.

## Checklist

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by nijkah** [comment]
Should I post a new PR for mmcv 2.x versino?

**Comment by zhouzaida** [comment]
> Should I post a new PR for mmcv 2.x versino?

Thanks for your contribution. I will cherry-pick this commit to 2.x.

File: mmcv/ops/multi_scale_deform_attn.py [hunk 1]
```diff
@@ -50,6 +50,17 @@ def forward(ctx, value: torch.Tensor, value_spatial_shapes: torch.Tensor,
...
 
         ctx.im2col_step = im2col_step
+
+        # When pytorch version >= 1.6.0, amp is adopted for fp16 mode;
+        # amp won't cast the type of sampling_locations, attention_weights
+        # (float32), but "value" is cast to float16, leading to the type
+        # mismatch with input (when it is float32) or weight.
+        # The flag for whether to use fp16 or amp is the type of "value",
+        # we cast weight and input to temporarily support fp16 and amp
+        # whatever the pytorch version is.
```
Comment (zhouzaida) [inline mmcv/ops/multi_scale_deform_attn.py:60]:
```suggestion
        # we cast sampling_locations and attention_weights to
        # temporarily support fp16 and amp whatever the
        # pytorch version is.
```

**Review by zhouzaida — APPROVED** [review]

**Comment by zhouzaida** [comment]
Please add unit tests to cover this modification. You can refer to https://github.com/open-mmlab/mmcv/blob/master/tests/test_ops/test_deform_conv.py.

**Comment by nijkah** [comment]
@zhouzaida Sorry for the late response. I added UT. 
But, I'm not sure about the precision threshold. Can I ask you to check it?

**Comment by zhouzaida** [comment]
> @zhouzaida Sorry for the late response. I added UT. But, I'm not sure about the precision threshold. Can I ask you to check it?

@nijkah Sorry for my late reply. The precision threshold is OK.

**Review by zhouzaida — APPROVED** [review]

**Pull request merged**